File: packages/types/src/environment.ts
/**
 * Definition of environmental information types
 *
 * Define type interfaces related to the runtime environment
 */

import type { Metadata } from "./common.js";

/**
 * Workspace Information
 */
export interface WorkspaceInfo {
  /**
   * Name of workspace
   */
  name: string;

  /**
   * Workspace path (absolute path)
   */
  path: string;

  /**
   * Workspace URI (optional)
   */
  uri?: string;

  /**
   * Is it a currently active workspace
   */
  isActive?: boolean;

  /**
   * Metadata (optional)
   */
  metadata?: Metadata;
}

/**
 * Environment Information Configuration
 */
export interface EnvironmentInfo {
  /**
   * Operating System Information
   */
  os: string;

  /**
   * Operating System Architecture
   */
  arch?: string;

  /**
   * Operating system version
   */
  osVersion?: string;

  /**
   * time zones
   */
  timezone: string;

  /**
   * user language
   */
  userLanguage: string;

  /**
   * User Area Settings
   */
  locale?: string;

  /**
   * Workspace paths (multiple workspaces supported)
   */
  workspaces: WorkspaceInfo[];

  /**
   * current timestamp
   */
  timestamp?: number;

  /**
   * Node.js version
   */
  nodeVersion?: string;

  /**
   * Additional environmental information
   */
  metadata?: Metadata;
}

/**
 * Environment Information Configuration Items
 */
export interface EnvironmentConfig {
  /**
   * Whether to include operating system details
   */
  includeOSDetails?: boolean;

  /**
   * Whether to include workspace information
   */
  includeWorkspaces?: boolean;

  /**
   * Whether to include the current time
   */
  includeTimestamp?: boolean;

  /**
   * Whether or not to include a Node.js version
   */
  includeNodeVersion?: boolean;

  /**
   * Customized workspace filters
   */
  workspaceFilter?: (workspace: WorkspaceInfo) => boolean;
}

File: crates/wf-types/src/environment.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WorkspaceInfo {
    pub name: String,
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uri: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_active: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<super::Metadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct EnvironmentInfo {
    pub os: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub arch: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub os_version: Option<String>,
    pub timezone: String,
    pub user_language: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub locale: Option<String>,
    pub workspaces: Vec<WorkspaceInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub node_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<super::Metadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct EnvironmentConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_os_details: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_workspaces: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_timestamp: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub include_node_version: Option<bool>,
}
